### PR TITLE
Fixed #33277 -- Disallowed database connections in threads in SimpleTestCase.

### PR DESCRIPTION
**Pull request by David-Wobrock**
[ticket-33277](https://code.djangoproject.com/ticket/33277)

Picked up from https://github.com/django/django/pull/17075

### CODE REVIEW

**Comment by felixxm** [comment]
@blueyed Does it work for you?

File: django/test/testcases.py [hunk 1]
```diff
@@ -266,6 +275,30 @@ def _remove_databases_failures(cls):
...
+                }
+                return _DatabaseFailure(self.ensure_connection, message)()
+
+            # Copied from BaseDatabaseWrapper.ensure_connection as the original as
+            # been patched at the base class level.
+            if self.connection is None:
+                with self.wrap_database_errors:
+                    self.connect()
+
+        return patched_ensure_connection
```
Comment (felixxm) [inline django/test/testcases.py:300]: I noticed that Sentry [is doing some interesting monkey patching](https://github.com/getsentry/sentry/blob/8c44dfaf66d8e534232b68817c94729f3248c360/src/sentry/testutils/cases.py#L550-L581) around `ensure_connection()`. Can this patch break it?

File: django/test/testcases.py [hunk 2]
```diff
@@ -266,6 +275,30 @@ def _remove_databases_failures(cls):
...
+                    "alias": self.alias,
+                    "operation": "threaded connections",
+                }
+                return _DatabaseFailure(self.ensure_connection, message)()
+
+            # Copied from BaseDatabaseWrapper.ensure_connection as the original as
+            # been patched at the base class level.
+            if self.connection is None:
+                with self.wrap_database_errors:
+                    self.connect()
```
Comment (felixxm) [inline django/test/testcases.py:298]: What about 3rd-party database backends with the custom `ensure_connection()` implementation?

Comment (felixxm) [reply django/test/testcases.py:298]: Also, the original `ensure_connection()` is `@async_unsafe` decorated.

Comment (David-Wobrock) [reply django/test/testcases.py]: Of course, copying the source code is a bad idea here 😛 
Let's call the actual `BaseDatabaseWrapper.ensure_connection` method.

Now, the monkey-patching approach means that, as long as we call `BaseDatabaseWrapper.ensure_connection`, the behaviour should work as expected.
Meaning that the Sentry implementation should keep working I believe.
But if another backend overrides `ensure_connection` without calling `super().ensure_connection(...)`, then the behaviour remains the same as today: database connections still work in threads.

Would there be another method(s) that we can leverage? 🤔

File: django/test/testcases.py (hunk 1 above)
Comment (David-Wobrock) [reply django/test/testcases.py:298]: With the latest pushed changes, I think that it should keep working 🤔 
See the comment on the other discussion: https://github.com/django/django/pull/17639#discussion_r1437704717

**Comment by David-Wobrock** [comment]
Thanks for the review @felixxm 🙏 I pushed some changes in the commit.

**Review by felixxm — COMMENTED** [review]
@David-Wobrock Thanks :+1:

I pushed small edits to a release note, and used `enterClassContext()`.

**Comment by devxpy** [comment]
Is there a workaround for old code relying on this bug (feature?)

**Pull request merged**